_sources/UsingRESTAPIs/cachingResponses.rst

```diff
@@ -15,120 +15,83 @@ You haven't experienced it yet, but if you get complicated data back from a REST
+    def params_unique_combination(baseurl, params_d, private_keys=["api_key"]):
+        alphabetized_keys = sorted(params_d.keys())
         res = []
         for k in alphabetized_keys:
-            res.append((k, d[k]))
-        return res
-
-    def requestURL(baseurl, params = {}):
-        req = requests.Request(method = 'GET', url = baseurl, params = canonical_order(params))
-        prepped = req.prepare()
-        return prepped.url
-
-    def get_with_caching(base_url, params_diction, cache_diction, cache_fname):
-        full_url = requestURL(base_url, params_diction)
-        # step 1
-        if full_url in cache_diction:
-            # step 2
-            print "retrieving cached result for " + full_url
-            return cache_diction[full_url]
+            if k not in private_keys:
+                res.append("{}-{}".format(k, params_d[k]))
+        return baseurl + "_".join(res)
+
+    # Function to get data from the datamuse API, using this caching pattern
+    def get_from_datamuse_caching(rhymes_with):
+        baseurl = "https://api.datamuse.com/words"
+        params_diction = {}
+        params_diction["rel_rhy"] = rhymes_with
+        unique_ident = params_unique_combination(baseurl,params_diction)
+        if unique_ident in CACHE_DICTION:
+            print("Getting cached data...")
+            return CACHE_DICTION[unique_ident]
         else:
-            # step 3
-            response = requests.get(base_url, params=params_diction)
-            print "adding cached result for " + full_url
-            # add to the cache and save it permanently
-            cache_diction[full_url] = response.text
-            fobj = open(cache_fname, "w")
-            pickle.dump(cache_diction, fobj)
-            fobj.close()
-            return response.text
-
-    dest_url = 'http://services.faa.gov/airport/status/DTW'
-    d = {'format': 'json'}
-    result_text = get_with_caching(dest_url, d, saved_cache, cache_fname)
-    print(json.loads(result_text))
\ No newline at end of file
+            print("Making a request for new data...")
+            # Make the request and cache the new data
+            resp = requests.get(baseurl, params_diction)
+            CACHE_DICTION[unique_ident] = json.loads(resp.text)
+            dumped_json_cache = json.dumps(CACHE_DICTION)
+            fw = open(CACHE_FNAME,"w")
+            fw.write(dumped_json_cache)
+            fw.close() # Close the open file
+            return CACHE_DICTION[unique_ident]
+
+
+In the next section, we'll break this code down, step by step.
+
+As an extra challenge, it's also worth noting that even supposing the exact same caching structure and setup, there are a number of different versions of that ``get_from_datamuse_caching`` function you could write. Can you think of some small ways to change the structure of that code so that the input and return value, and the data saved in the cache file, would all remain exactly the same?
```

Comment:
I'll accept this as is, but I think it really doesn't make sense as written. Do you want them to change the keys that are used for doing lookups in the cache? Or do you really mean that the cache file will be exactly the same and they are just rewriting some of the code? Part of the idea of modularity is that they should be able to change the internals (not just the code but also the structure of the file) as long as they keep the inputs and outputs the same.